Add update docs for rhel edition (#192)
RotherOSS/doc-otobo-installation: content/updating/updating-rhel.rst

Updating

> ========

Reviewer: ```suggestion
Updating on RHEL
================
```

> This document describes the procedure for upgrading an existing OTOBO 11.0 to 11.1.

Reviewer: ```suggestion
This document describes the procedure for upgrading an existing OTOBO 11.0 to 11.1 on RHEL.
For other platforms, refer to the other guides :doc:`../updating`
```

However, the same procedure can be used for patch level updates, e.g., 11.1.0 to 11.1.1.
Just omit the steps not relevant for patch level updates.

.. note::

   It is highly recommended to perform a test update on a separate testing machine first.

.. note::

   To prepare required Perl modules beforehand, you can already download and unzip the new OTOBO version and execute the check modules script, e.g. with the following command (please adjust the file path).
   This is not necessary in a regular installation and will be part of the later update instruction.

   .. code-block:: bash

>         root> export PERL5LIB="/opt/otobo/install/local/lib/perl5"

Reviewer: Could this look like this?

```suggestion
        sudo env \
        PERL5LIB="/opt/otobo/install/local/lib/perl5" \
        perl /opt/otobo/bin/otobo.CheckModules.pl --list
```

Using `root> ` means you cannot copy the block with the button.

        root> perl /opt/otobo/bin/otobo.CheckModules.pl --list

.. note::

>    On Ubuntu/Debian systems you have to manually install some Perl packages before upgrading to 11.1.

Reviewer: Why is this part of rhel?

        root> dnf install -y wget perl perl-DBD-MySQL libpq-devel libxslt-devel libxml2-devel graphviz-devel unixODBC-devel xz-devel

For OTOBO 11.1, the following packages are being migrated automatically to the framework.
This means that no separate package is necessary and they will be part of OTOBO by default.

   - CK5-FullWindowMode
   - CustomerAgeShowCreated
   - CustomerTicketSearch
   - Elasticsearch-Extension
   - ExtendedArticleEdit
   - HideShowForAgentTicketCompose
   - ImportExportCustomerCompany
   - ImportExportStandardObjects
   - ImportExportTicket
   - PostMasterXFromHeader
   - ProcessTicketTemplates
   - RestorePendingInformation
   - RotherOSS-AccountedTimeInViews
   - TicketUpdateOperationExternalIdentifier
   - OAuth2
   - OAuth2-Mail
   - Elasticsearch-FAQ

.. warning::
   The optional 11.0 package 'MailAccount-OAuth2' is obsolete and replaced by new functionality in OTOBO core.
   It will not be uninstalled during migration to OTOBO 11.1.
   It will be listed as not fully installed.
   This will allow you to migrate configuration from the old package to new core functionality.
   You need to uninstall the package manually after migration.

> 

Reviewer: This should go from every specific guide into the generic preface.

Step 1: Stop All Relevant Services and the OTOBO Daemon
-------------------------------------------------------

Please make sure there are no more running services or Cron jobs that try to access OTOBO.
This will depend on your service configuration.

.. code-block:: bash

   root> systemctl stop postfix
   root> systemctl stop nginx

>    root> systemctl stop cron

Reviewer: Why stop `cron`? The install instructions do not mention it.

Author: No, but this has been recommended in every other updating doc too. Even though it is configured nowhere.

Reviewer: That is probably because when not using systemd, cron is used to bring the daemon up. But with RHEL, we never used cron.

Now you need to be stopping the OTOBO core services.

.. code-block:: bash

    root> systemctl disable --now otobo-web.service otobo-daemon.service

Step 2: Backup Files and Database
---------------------------------

> Create a backup of the hole ``/opt/otobo`` directory and the database.

Reviewer: ```suggestion
Create a backup of the whole ``/opt/otobo`` directory and the database.
```

> Example for a Standard Installation with Ubuntu and MySQL

Reviewer: 1. Ubuntu?
2. Why not rely on backup and restore? If backup and restore are lacking in any way, we should know.

Author: This is the default way when upgrading. This was discussed with Bernhard and was already part of every other upgrading documentation.

Reviewer: Again. We should know why that is and not maintain two procedures for the same thing.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: bash

    root> mkdir /opt/otobo-update                            # Create a update directory
    root> cd /opt/otobo-update                               # Change into the update directory
    root> cp -pr /opt/otobo otobo-prod-old                   # Backup the whole OTOBO directory into the update directory
    root> mysqldump -u otobo -p otobo -r otobo-prod-old.sql  # Backup the otobo database to otobo-prod-old.sql

Please check whether all files are valid.

.. warning::

    Do not proceed without a complete backup of your system.

>     You can also use the :doc:`/content/backup-restore` script for this.

Reviewer: Ref probably broken.

Step 3: Install the new Release
-------------------------------

Download the latest OTOBO release from https://ftp.otobo.org/pub/otobo/ and unpack the source archive (for example, using ``tar``) into the directory ``/opt/otobo-update``:

.. code-block:: bash

    root> cd /opt/otobo-update                                              # Change into the update directory
    root> wget https://ftp.otobo.org/pub/otobo/otobo-11.1.0-beta1.tar.gz    # Download the latest OTOBO 11.1 release
    root> tar -xzf otobo-11.1.0-beta1.tar.gz                                # Unzip OTOBO
    root> \cp -r otobo-11.1.*/* /opt/otobo                                  # Copy the new otobo directory to /opt/otobo

Restore Old Configuration Files
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> We need only copy the file ``Kernel/Config.pm`` in OTOBO 11.

Reviewer: Sentence broken.

.. code-block:: bash

    root> cd /opt/otobo-update

>     root> \cp -p otobo-prod-old/Kernel/Config.pm /opt/otobo/Kernel/

Reviewer: Why are there `\`?

Author: Because if not the cp module in RHEL asks on every file if you want to overwrite it or not. This is especially anoying when trying to copy the folder from the new version.

Reviewer: That is probably because `cp` is an alias to `cp -i`. You should check and give an explanation somewhere.

    root> \cp -p otobo-prod-old/var/cron/* /opt/otobo/var/cron/

Restore Article Data
~~~~~~~~~~~~~~~~~~~~

> If you configured OTOBO to store article data in the file system, restore the ``article`` folder to ``/opt/otobo/var/`` or the folder specified in the system configuration.

Reviewer: This is a setting. Please mark the assumption.

.. code-block:: bash

    root> cd /opt/otobo-update

>     root> \cp -pr otobo-prod-old/var/article/* /opt/otobo/var/article/

Reviewer: Why `\`?

Author: See above

Reviewer: see above.

Restore Already Installed Default Statistics
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> If you have additional packages with default statistics, restore the stats XML files with the suffix ``*.installed`` to ``/opt/otobo/var/stats``.

Reviewer: ```suggestion
If you have packages installed, which provide additional statistics, restore the XML files with the suffix ``*.installed`` to ``/opt/otobo/var/stats``:
```

.. code-block:: bash

    root> cd /opt/otobo-update/otobo-prod-old/var/stats
    root> \cp *.installed /opt/otobo/var/stats

Set File Permissions
~~~~~~~~~~~~~~~~~~~~

Execute the following command to set the file and directory permissions for OTOBO.
It will try to detect the correct user and group settings needed for your setup.

.. code-block:: bash

    root> export PERL5LIB="/opt/otobo/install/local/lib/perl5"
    root> /opt/otobo/bin/otobo.SetPermissions.pl --otobo-user=otobo --web-group=otobo
    root> chmod +x /opt/otobo/install/local/bin/*

Check Webserver Configuration Files
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Newer versions of OTOBO may need you to adjust the webserver configuration.

Reviewer: 1. This is not very descriptive. 
2. This is the same for the native installation for Ubuntu and RHEL. Describe it once, reference from both places. Maybe a how-to or the updating preface?

Author: I would rather completely remove this section from the rhel documentation because RHEL systems are already created with a up-to-date config according to the install manual.
Maybe we can add a section for this in updating.rst for future releases when there are required changes.

Reviewer: Agreed.

We provide Nginx templates at ``scripts/nginx-vhost-443.include.conf`` and ``scripts/nginx-vhost-80.include.conf``.

Step 4: Install new needed Perl Modules
-----------------------------------------

> OTOBO needs new cpan packages for some version jumps.

Reviewer: ```suggestion
OTOBO needs new CPAN packages to allign the installed versions with OTOBOs requierements.
```
